core/src/Streamly/Internal/Unicode/Stream.hs

```diff
@@ -682,14 +700,254 @@ decodeUtf8Lax = decodeUtf8
+scanW16FromEvenChunks
+    :: Monad m
+    => ByteOrdering -- ^ Byte orderdering of the machine where the value is
+                    -- being deserialized.
+    -> ByteOrdering -- ^ Byte orderdering of the machine where the value is
+                    -- serialized.
+    -> Stream m (Maybe (Word8, Word8), Array Word8)
+    -> Stream m Word16
+scanW16FromEvenChunks dBO sBO (D.Stream step state) =
+    D.Stream step1 (SFCOuter state)
+
+    where
+
+    hasSameByteOrdering = dBO == sBO
+
+    {-# INLINE_LATE step1 #-}
+    step1 gst (SFCOuter st) = do
+        r <- step (adaptState gst) st
+        return $
+            case r of
+                Yield (Nothing, Array {..}) st1 ->
+                    Skip (SFCInner st1 arrContents arrStart arrEnd)
+                Yield (Just (w1, w2), Array {..}) st1 ->
+                    let w1As16 = fromIntegral w1
+                        w2As16 = fromIntegral w2
+                     in
+                       case sBO of
+                          LittleEndian ->
+                              Yield
+                                 ((w2As16 `shiftL` 8) .|. w1As16)
+                                 (SFCInner st1 arrContents arrStart arrEnd)
+                          BigEndian ->
+                              Yield
+                                 ((w1As16 `shiftL` 8) .|. w2As16)
+                                 (SFCInner st1 arrContents arrStart arrEnd)
+                Skip st1 -> Skip (SFCOuter st1)
+                Stop -> Stop
+
+    step1 _ (SFCInner st _ i end) | i == end = pure $ Skip $ SFCOuter st
+    step1 _ (SFCInner st arrC i end) =
+        pure $
+            if hasSameByteOrdering
+            then
+                let !v = unsafeInlineIO $ Unbox.peekAt i arrC
+                 in Yield v (SFCInner st arrC (i + 2) end)
+            else
+                let !w1 = unsafeInlineIO $ Unbox.peekAt i arrC
+                    !w2 = unsafeInlineIO $ Unbox.peekAt (i + 1) arrC
+                    w1As16 = w8Tow16 w1
+                    w2As16 = w8Tow16 w2
+                 in
+                   case sBO of
+                       LittleEndian ->
+                           let v = (w2As16 `shiftL` 8) .|. w1As16
+                            in Yield v (SFCInner st arrC (i + 2) end)
+                       BigEndian ->
+                           let v = (w1As16 `shiftL` 8) .|. w2As16
+                            in Yield v (SFCInner st arrC (i + 2) end)
+
+{-# INLINE scanW16FromChunks #-}
+scanW16FromChunks
+    :: Monad m
+    => ByteOrdering -- ^ Byte orderdering of the machine where the value is
+                    -- being deserialized.
+    -> ByteOrdering -- ^ Byte orderdering of the machine where the value is
+                    -- serialized.
+    -> Stream m (Array Word8)
+    -> Stream m Word16
+scanW16FromChunks dBO sBO strm =
+    scanW16FromEvenChunks dBO sBO (mkEvenW8Chunks strm)
+
+-- | Create a Word16 Stream from Word8 chunks with little endian byte ordering.
```

Comment:
Create a Word16 stream from Word8 chunks interpreted in little-endian byte ordering.